src/plan/generational/copying/global.rs

```diff
@@ -138,7 +138,12 @@ impl<VM: VMBinding> Plan for GenCopy<VM> {
     /// Return the number of pages avilable for allocation. Assuming all future allocations goes to nursery.
     fn get_pages_avail(&self) -> usize {
         // super.get_pages_avail() / 2 to reserve pages for copying
-        (self.get_total_pages() - self.get_pages_reserved()) >> 1
```

Comment:
I noticed this causing an issue with some test cases. It might be an issue for the [default implementation](https://github.com/mmtk/mmtk-core/blob/master/src/plan/global.rs#L264-L266) as well, but we've never triggered it.

Reply:
The total pages should not be greater than or equal to reserved pages. If this panics, probably there is a bug somewhere else about page accounting. I don't suggest simply using `saturating_sub`.

Reply:
Is that true? It seems like `collection_required()` expects it to hold sometimes: https://github.com/mmtk/mmtk-core/blob/master/src/plan/global.rs#L784

Reply:
Ah, you are right. For failed allocation (OOM), it could happen. We will increase reserved pages temporarily, and check if that exceeds total pages. If so, we will deduct the temporary reserved pages and trigger a GC.

Did you fix this because it panics here with changes in this PR?

Reply:
I honestly didn't try with master. I just noticed GenCopy failing in debug builds due to subtraction overflow while I was testing my implementation. I will check if GenCopy fails with master.

Reply:
It does fail on master.

Reply:
What did you run to trigger the bug? Can you post the command line?

Reply:
```
RUST_BACKTRACE=1 RUST_LOG= MMTK_PLAN=GenCopy ./build/linux-x86_64-normal-server-fastdebug/jdk/bin/java -XX:+UseThirdPartyHeap -Xms5M -Xmx5M -jar /usr/share/benchmarks/dacapo/dacapo-evaluation-git-29a657f.jar zxing
--------------------------------------------------------------------------------
IMPORTANT NOTICE:  This is NOT a release build of the DaCapo suite.
Since it is not an official release of the DaCapo suite, care must be taken when
using the suite, and any use of the build must be sure to note that it is not an
offical release, and should note the relevant git hash.

Feedback is greatly appreciated.   The preferred mode of feedback is via github.
Please use our github page to create an issue or a pull request.
    https://github.com/dacapobench/dacapobench.
--------------------------------------------------------------------------------

thread '<unnamed>' panicked at 'attempt to subtract with overflow', /home/kunals/git/mmtk-core/src/plan/generational/copying/global.rs:141:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/panicking.rs:498:5
   1: core::panicking::panic_fmt
             at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/core/src/panicking.rs:107:14
   2: core::panicking::panic
             at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/core/src/panicking.rs:48:5
   3: <mmtk::plan::generational::copying::global::GenCopy<VM> as mmtk::plan::global::Plan>::get_pages_avail
             at /home/kunals/git/mmtk-core/src/plan/generational/copying/global.rs:141:9
   4: mmtk::plan::generational::global::Gen<VM>::should_next_gc_be_full_heap
             at /home/kunals/git/mmtk-core/src/plan/generational/global.rs:217:9
   5: <mmtk::plan::generational::copying::global::GenCopy<VM> as mmtk::plan::global::Plan>::release
             at /home/kunals/git/mmtk-core/src/plan/generational/copying/global.rs:127:36
   6: <mmtk::scheduler::gc_work::Release<C> as mmtk::scheduler::work::GCWork<<C as mmtk::scheduler::work::GCWorkContext>::VM>>::do_work
             at /home/kunals/git/mmtk-core/src/scheduler/gc_work.rs:114:9
   7: mmtk::scheduler::work::GCWork::do_work_with_stat
             at /home/kunals/git/mmtk-core/src/scheduler/work.rs:37:9
   8: mmtk::scheduler::worker::GCWorker<VM>::run
             at /home/kunals/git/mmtk-core/src/scheduler/worker.rs:132:13
   9: mmtk::memory_manager::start_worker
             at /home/kunals/git/mmtk-core/src/memory_manager.rs:177:5
  10: start_worker
             at /home/kunals/git/mmtk-openjdk/mmtk/src/api.rs:124:5
  11: _ZN6Thread8call_runEv
             at ./src/hotspot/share/runtime/thread.cpp:385
  12: thread_native_entry
             at ./src/hotspot/os/linux/os_linux.cpp:787
  13: start_thread
             at /build/glibc-S9d2JN/glibc-2.27/nptl/pthread_create.c:463
  14: __GI___clone
             at /build/glibc-S9d2JN/glibc-2.27/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:95
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
fatal runtime error: failed to initiate panic, error 5
zsh: abort (core dumped)  RUST_BACKTRACE=1 RUST_LOG= MMTK_PLAN=GenCopy  -XX:+UseThirdPartyHeap -Xms5M
```

Reply:
It seems there is a bug in our page accounting that after a GC, we end up with more used pages than before the GC. I will look further into this.

Reply:
Should I remove this change then? It only hides the issue instead of fixes it.

Reply:
I think this is actually possible (see: https://mmtk.zulipchat.com/#narrow/stream/262673-mmtk-core/topic/copy-space-pages-used/near/270335481). I think the fix is right, and we possibly want to fix all of the occurrences of `total_pages - reserved pages`, and make it clear that it is possible that the reserved pages is larger than the total pages.

Reply:
Right. Should I fix the other ones in this PR as well or will you create a new PR for the changes?

Reply:
Either sounds good to me. Making it in a separate PR is cleaner, and we can quickly get it merged even before this PR.

Reply:
I agree. I'll remove the change from this PR then.

Reply:
Done. I think it would be best if you made the PR fixing the above since you've debugged it and understand the issue better than me. I'll mark this conversation as resolved. Thanks for looking into it!